File: 1044-find-common-characters/1044-find-common-characters.cpp
class Solution {
public:
    vector<string> commonChars(vector<string>& words) {
        int n = words.size();
        vector<int> commonCharCnt(26, 0), currentCharCnt(26, 0);
        for(char c : words[0]) commonCharCnt[c-'a']++;
        for(int i=1; i<words.size(); i++){
            fill(currentCharCnt.begin(), currentCharCnt.end(), 0);
            //Count char in the current word
            for(char c : words[i]) currentCharCnt[c-'a']++;
            //Update the common char count to keep the minimum counts
            for(int j=0; j<26; j++){
                commonCharCnt[j] = min(commonCharCnt[j], currentCharCnt[j]);
            }
        }
        vector<string> ans;
        //COllect the common char to keep the minimum counts
        for(int i=0; i<26; i++){
            for(int j=0; j<commonCharCnt[i]; j++) ans.push_back(string(1, i+'a'));
        }
        return ans;
    }
};

File: 1044-find-common-characters/1044-find-common-characters.java
class Solution {
    public List<String> commonChars(String[] words) {
        int n = words.length;
        int[] commonCharCnt = new int[26];
        int[] currentCharCnt = new int[26];
        for(char c : words[0].toCharArray()) commonCharCnt[c-'a']++;
        for(int i=1; i<n; i++){
            Arrays.fill(currentCharCnt, 0);
            for(char c : words[i].toCharArray()) currentCharCnt[c-'a']++;
            for(int j=0; j<26; j++){
                commonCharCnt[j] = Math.min(commonCharCnt[j], currentCharCnt[j]);
            }
        }
        List<String> ans = new ArrayList<>();
        for(int i=0; i<26; i++){
            for(int j=0; j<commonCharCnt[i]; j++) ans.add(String.valueOf((char)(i+'a')));
        }
        return ans;
    }
}